### PR TITLE
Make `-preview=in` means `[ref] const scope` and accept rvalues

### PR DESCRIPTION
**Pull request by Geod24**
```
As explained in the changelog entry, this puts `in` more in line with its meaning:
that of an input parameter.
This change was prompted by the introduction of `-preview=in` to make `in` means
`const scope` (its original meaning). If we introduce a `preview` switch,
we might as well go all the way.
```

Needs test, I know, and won't pass the CI because druntime has scope `const scope` inheriting `in`.
But before I dig into making this work, I'd like to start the discussion on the concept itself.

The goal is to make `in` the "go to" type for input parameter. I had this in mind for a long time, and discussed it with @JinShil on some occasions.
Now that `-preview=in` has been merged ( #10769 ), I figured it would be the right time to introduce it, hoping to get his approved before the next release.

CC @atilaneves @WalterBright

### CODE REVIEW

**Comment by kinke** [comment]
I think the optimal solution would be either `const scope` or `const scope ref`, depending on the type and target ABI, for the most efficient way to pass an input parameter. I.e., no `ref` (extra indirection) for class references and small PODs (incl. delegates and slices for most ABIs), but a ref for non-PODs and large PODs.

**Comment by Geod24** [comment]
@kinke : I agree. However, why limit it to `in` ? Could we make it an ABI thing that small types can be copied instead of passed by `const ref` when possible ?

**Comment by atilaneves** [comment]
I think this takes too much control away from the programmer. What if I don't *want* to pass by ref? Sure, I could declare `const scope` instead of `in`, but given that in D move semantics are achieved with pass-by-value, is `ref` something we want to encourage?

**Comment by kinke** [comment]
> but given that in D move semantics are achieved with pass-by-value, is ref something we want to encourage?

(Optimized) DMD codegen: [godbolt](https://d.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(fontScale:14,j:1,lang:___d,selection:(endColumn:2,endLineNumber:26,positionColumn:2,positionLineNumber:26,selectionStartColumn:2,selectionStartLineNumber:26,startColumn:2,startLineNumber:26),source:'struct+S%0A%7B%0A++++long%5B4%5D+data%3B%0A%7D%0A%0Avoid+byVal(S)%3B%0Avoid+byRef(ref+S)%3B%0A%0Avoid+copyByVal()%0A%7B%0A++++S+lval%3B%0A++++byVal(lval)%3B%0A%7D%0A%0Avoid+moveByVal()%0A%7B%0A++++import+core.lifetime+:+move%3B%0A++++S+lval%3B%0A++++byVal(move(lval))%3B%0A%7D%0A%0Avoid+passByRef()%0A%7B%0A++++S+lval%3B%0A++++byRef(lval)%3B%0A%7D'),l:'5',n:'0',o:'D+source+%231',t:'0')),k:50.40371890861558,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:dmd2nightly,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'1',trim:'1'),fontScale:14,j:1,lang:___d,libs:!(),options:'-O+-inline',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'dmd+nightly+(Editor+%231,+Compiler+%231)+D',t:'0')),k:49.59628109138442,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)
LDC: [godbolt](https://d.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(fontScale:14,j:1,lang:___d,selection:(endColumn:2,endLineNumber:26,positionColumn:2,positionLineNumber:26,selectionStartColumn:2,selectionStartLineNumber:26,startColumn:2,startLineNumber:26),source:'struct+S%0A%7B%0A++++long%5B4%5D+data%3B%0A%7D%0A%0Avoid+byVal(S)%3B%0Avoid+byRef(ref+S)%3B%0A%0Avoid+copyByVal()%0A%7B%0A++++S+lval%3B%0A++++byVal(lval)%3B%0A%7D%0A%0Avoid+moveByVal()%0A%7B%0A++++import+core.lifetime+:+move%3B%0A++++S+lval%3B%0A++++byVal(move(lval))%3B%0A%7D%0A%0Avoid+passByRef()%0A%7B%0A++++S+lval%3B%0A++++byRef(lval)%3B%0A%7D'),l:'5',n:'0',o:'D+source+%231',t:'0')),k:50.40371890861558,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:ldclatestci,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'1',trim:'1'),fontScale:14,j:1,lang:___d,libs:!(),options:'-O',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'ldc+latest+CI+(Editor+%231,+Compiler+%231)+D',t:'0')),k:49.59628109138442,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

**Comment by kinke** [comment]
> However, why limit it to in ? Could we make it an ABI thing that small types can be copied instead of passed by const ref when possible ?

You mean `const ref T param` could suddenly mean `const T param` for small PODs, and so `&param` yield something completely different? That won't work IMO.

Limited to `in` as a way to make this potential optimization transparent, and to give the user the possibility to check whether it's a ref or not via `__traits(isRef)` if need be.

**Comment by benjones** [comment]
Not sure if this is the best place to leave this, but Herb Sutter's preliminary proposal seems related.  Basically, he says the programmer should specify intent (read only, output parameter, etc) and the compiler should choose whether things are passed by pointer or value and ensure legal usage: https://www.youtube.com/watch?v=qx22oxlQmKc&t=3454s

**Comment by Geod24** [comment]
@kinke : Good point, I didn't think about taking the address of it. And things like fiber context switch would make that visible to the user, even if `scope` is involved. So yes, having it selectively applied sounds better.

> I think this takes too much control away from the programmer.

It doesn't take *any* control away. One is still free to apply the storage classes as one see fit. It doesn't restrict what can be done.

However, it makes it easy to do the right thing by default. You have an input parameter, you put `in` on it, and you're done. It's self-documenting and it doesn't make you think about the details of what's happening under the hood. It's all about making things easy in the common case.

Regarding move semantic, what @kinke said. The code generated by DMD is terrible, and uncontrollable (you can't prevent a struct being moved). Over the past two years, we've seen work in that direction:
- [The accepted copy constructor DIP](https://github.com/dlang/DIPs/blob/master/DIPs/accepted/DIP1018.md)
- [The accepted move constructor DIP](https://github.com/dlang/DIPs/blob/master/DIPs/accepted/DIP1014.md)
- [Walter's pending DIP](https://github.com/dlang/DIPs/pull/182)

The abstract of that last link directly mention copy elision as a goal for the DIP.

**Comment by Geod24** [comment]
Tests with my projects pass. Going to extend the usage of `in` and see if it holds, but this is definitely ready for review.

**Comment by UplinkCoder** [comment]
Wow this is a big one.

We should have left a Prefix character for mangle extensions ....

**Comment by Geod24** [comment]
Alright, the `in` part was merged, so there's that already.

I still have a few things to tackle:
- Need to add CTFE support (it triggers on Phobos);
- There seem to be an issue with template deduction (although that might be due to `scope`) in Mir;
- Using the type's `size` means that forward references error can be triggered earlier, as seen in `ikod/dlang-requests`;
- Mir has a few "goto skips over" error messages - those might be legit, or not;
- For most project to be properly tested, I'll need to add a workaround in taggedalgebraic to detect if `in` can be `ref`;
- A few projects in Buildkite just have scope errors, so those needs to be fixed;

Obviously I don't intend to have this merged with the commit that makes it the default, but if it passes Buildkite, we can be pretty confident it works.

**Comment by anon17** [comment]
> it is only passed by `ref` if it cannot be passed in register.

You mean if it fits in two registers, it's still passed by reference? 16 bytes is the size of uuid and is better passed by value.

**Comment by tsbockman** [comment]
> You mean if it fits in two registers, it's still passed by reference? 16 bytes is the size of uuid and is better passed by value.

Also, `__vector` types and `struct`s that transitively wrap a single `__vector` member are also generally passed by register.

Also, whether a specific parameter can be passed by register depends upon how many registers have already been consumed by other parameters; there is a platform-dependent limit on how many registers of a given type (on x86, general purpose and floating-point/SIMD registers are counted separately) can be consumed by parameters.

**Comment by anon17** [comment]
I think consumed registers don't affect it? Whether it's passed by value or by reference shouldn't depend on its position.

**Comment by tsbockman** [comment]
> I think consumed registers don't affect it? Whether it's passed by value or by reference shouldn't depend on its position.

Here's an example of what I mean. This D code:
```D
auto passByTest(const(long[2]) arg0, const(long[2]) arg1, const(long[2]) arg2, const(long[2]) arg3, const(long[2]) arg4) {
    long ret = 0;
    ret += arg0[0];
    ret += arg0[1];
    ret += arg1[0];
    ret += arg1[1];
    ret += arg2[0];
    ret += arg2[1];
    ret += arg3[0];
    ret += arg3[1];
    ret += arg4[0];
    ret += arg4[1];
    return ret;
}
```
Compiles to this assembly code on x86-64 using LDC:
```asm
pure nothrow @nogc @safe long glycogen.math.passByTest(const(long[2]), const(long[2]), const(long[2]), const(long[2]), const(long[2])):
        lea     rax, [rdi + rsi]
        add     rax, rdx
        add     rax, rcx
        add     rax, r8
        add     rax, r9
        add     rax, qword ptr [rsp + 8]
        add     rax, qword ptr [rsp + 16]
        add     rax, qword ptr [rsp + 32]
        add     rax, qword ptr [rsp + 24]
        ret
```
Note that although all parameters have the same type, `const(long[2])`, the first three parameters are received in the registers `rdi`, `rsi`, `rdx`, `rcx`, `r8`, and `r9`, but the last three are on the stack at `[rsp + 8]` and `[rsp + 32]`.

It *must* work this way, because there are a finite number of registers, and a function may have more parameters than there are registers of the appropriate type. (Additionally, some registers are reserved for other purposes.)

**Comment by kinke** [comment]
Yeah, but in that case it's still beneficial to pass the 2 spilled params by value - resulting in 2x2 longs on the stack instead of 2 pointers on the stack.

The backend has all the info and knows how to best pass an argument for the target ABI, so there should probably be a hook in the `Target` struct to decide whether a POD is to be passed by value or ref. The frontend can enforce a ref for non-PODs.

**Comment by Geod24** [comment]
Working on moving the hook to `target.d` after the basic frontend checks for postblit / dtor / copy ctor.

Otherwise, blocked by the following bugs:
- https://issues.dlang.org/show_bug.cgi?id=20179
- https://issues.dlang.org/show_bug.cgi?id=17351

**Comment by nordlow** [comment]
> The backend has all the info and knows how to best pass an argument for the target ABI (e.g.: PODs > 64 bit must be passed by ref for the Win64 ABI), 

What should the default be for Linux? Can I make use of such branching on ref or value passing right now in my code depending on size of type?

**Comment by Geod24** [comment]
> Can I make use of such branching on ref or value passing right now in my code depending on size of type?

I'm not sure what your question is ? Currently the PR is still here (as mentioned, I need to fix other bugs before we merge this), so no you can't use it.

**Comment by kinke** [comment]
For Posix x86_64 targets, the DMD implementation here should probably use the [argTypes machinery](https://github.com/dlang/dmd/blob/fa4f32ea95ce257a881bdafe2e8ae8f20d2139b1/src/dmd/argtypes_sysv_x64.d#L31) (cached [here](https://github.com/dlang/dmd/blob/457074e8ea12a56d9db9a8edb9427e21e15ee94f/src/dmd/dstruct.d#L226)) - a non-null and non-empty tuple meaning it should be passed by value. [These argTypes are accessible by user code too.]

For Win64, the rule is pretty simple - only PODs <= 64 bits and of a power-of-2 size are passed in registers, with exceptions for vectors and vector-wrapping structs IIRC. So e.g. `ubyte[3]` and (DMD 80-bit) `real` should be passed by ref.

For 32-bit x86 targets, the D ABI only uses a single register for register passing, so we could probably go with the less-than-or-equal-to-2-machine-words heuristic.

**Comment by UplinkCoder** [comment]
@Geod24 actually while we are on this.
Can we reserve I for extensions please.
so can you do your mangle prefix be `Ia` or something.

**Comment by rainers** [comment]
> Can we reserve I for extensions please.

There are already "extension manglings" beginning with 'N' and 'z'.

**Comment by nordlow** [comment]
> For Posix x86_64 targets, the DMD implementation here should probably use the [argTypes machinery](https://github.com/dlang/dmd/blob/fa4f32ea95ce257a881bdafe2e8ae8f20d2139b1/src/dmd/argtypes_sysv_x64.d#L31) (cached [here](https://github.com/dlang/dmd/blob/457074e8ea12a56d9db9a8edb9427e21e15ee94f/src/dmd/dstruct.d#L226)) - a non-null and non-empty tuple meaning it should be passed by value. [These argTypes are accessible by user code too.]

Are you saying I can look up preferred passing mechanism from `argTypes` via some public builtin API?

**Comment by nordlow** [comment]
What to do about

```
fail_compilation/warn7444.d(22): Error: cannot implicitly convert expression `e` of type `int` to `int[]`
fail_compilation/warn7444.d(26): Warning: assignment of `sa` to itself has no side effect
fail_compilation/warn7444.d(42): Warning: assignment of `da` to itself has no side effect
```

?

**Comment by Geod24** [comment]
A few questions that arose:

- Should  `in ref` really be allowed ? Currently it's a pattern that is used (e.g. [in Phobos](https://github.com/dlang/phobos/blob/e36b5aaba05444ce89197df1b2d075f08fff14af/std/uuid.d#L248-L261)).
  The only reason you'd want to use `in ref` is to *force* something to be `ref`. But then, what's the point of using `in` ?
  The main advantage you'd have by using `in ref` over `scope const ref` is that the former will accept rvalues.
  The reason I'm asking is that in multiple places, this creates dangerous situations where two functions are emitted, and the call is ambiguous.

- There are a few places where we have runtime hooks using `in`. Unfortunately, those runtime hooks usually don't go through semantic, so they end up not having their type properly set up (it's done manually at the invocation site). Those will need manual adjustment / to be kept in sync,  which is obviously a terrible thing to do. Luckily, there's no place which would have promoted a value to a ref (or I haven't found it yet). I am thinking of modifying those that I found (e.g. AA) so that they go through the proper semantic pass.

- What about Phobos ? If we provide the switch, we'll end up with the same issue as with `scope`: Different ABI for different switches. In the case of `scope` it was mangling so pretty easy to spot, but for this, it'll be ABI. That's not going to be nice for the user.

- Likewise, debugger experience: How will someone inspecting a function know if the parameters are passed by value or by ref ? One will need to inspect the function prologue.

As a result, I am seriously considering disallowing `in ref`. I could then just use it to make sure we never have a mangled function whose ABI is not clear. Thoughts @kinke @ibuclaw @rainers ?

**Comment by kinke** [comment]
> What about adding this as a trait in druntime?

These argTypes are (currently) only used for Posix x64 (and AArch64 for LDC), and not used by GDC at all AFAIK.

**Comment by kinke** [comment]
> Should `in ref` really be allowed ? Currently it's a pattern that is used (e.g. in Phobos).
> The only reason you'd want to use in ref is to *force* something to be ref. But then, what's the point of using `in` ?
> The main advantage you'd have by using `in ref` over `scope const ref` is that the former will accept rvalues.

I think it should ideally still be allowed - simply because it's nicer to type than `const scope ref` in (the probably rare) case one is interested in an lvalue ref. A 2nd overload using only `in` should be ambiguous and fail to compile.
If the ambiguity check significantly complicates matters, I personally wouldn't miss `in ref` a lot.

> There are a few places where we have runtime hooks using in. [...] I am thinking of modifying those that I found (e.g. AA) so that they go through the proper semantic pass.

Hmm, I guess those shouldn't use any `in` indeed and be fully specified instead.

> What about Phobos ? [...] In the case of scope it was mangling so pretty easy to spot, but for this, it'll be ABI.

I haven't really looked at the code diff, but assumed you wouldn't touch the mangling and simply lower `in` to `const scope [ref]`, so that ref and value should mangle differently.

> Likewise, debugger experience: How will someone inspecting a function know if the parameters are passed by value or by ref ?

Ref params are declared as DWARF refs, at least for LDC, and I assume that info can be queried by the debugger.

**Comment by Geod24** [comment]
> I haven't really looked at the code diff, but assumed you wouldn't touch the mangling and simply lower `in` to `const scope [ref]`, so that `ref` and value should mangle differently.

Oh... No, I touched the mangling *before* this PR so that `in` would be mangled as its own storage class.

> Ref params are declared as DWARF refs, at least for LDC, and I assume that info can be queried by the debugger.

Right, and I suppose Windows debug infos have a similar property ?

**Comment by kinke** [comment]
> No, I touched the mangling before this PR so that in would be mangled as its own storage class.

I guess there's a good reason (header generation, error msgs, ...), but it complicates matters for Phobos as you've mentioned. I don't think we have many options - druntime and Phobos should probably be compiled according to the default setting, and alternative CLI settings will require re-compiling the libs to prevent segfaults and the like.

> Right, and I suppose Windows debug infos have a similar property ?

Yes, just tested with LDC - a `ref int` param showing up as `int*` (`-g`) or `int&` (`-gc`) in the VS debugger with Visual D installed. [I've only mentioned DWARF because that's what LDC generates, LLVM then converts that to CodeView if needed.]

**Comment by Geod24** [comment]
Another alternative is to just detect it: https://github.com/s-ludwig/taggedalgebraic/pull/59

**Comment by anon17** [comment]
> Should in ref really be allowed ?

Yes? Otherwise how to pass an input argument by reference? Also don't forget your option is optin. The reasons you presented don't look like they conflict with `in ref`. The point of `in` is to specify `scope const` parameter. Runtime hooks can't possibly not go through semantic, because there's no other way to get to codegen stage?

**Comment by Geod24** [comment]
> Yes? Otherwise how to pass an input argument by reference?

You let the compiler decide. There's no reason to force `ref` on `in` parameter if the compiler thinks you don't need it.

> Runtime hooks can't possibly not go through semantic, because there's no other way to get to codegen stage?

Well they don't, because they are `extern (C)` functions that get compiled in druntime and DMD just emits a call.

**Comment by Geod24** [comment]
I just pushed what should be the final version.
The changelog was rewritten to explain the full rationale of the change.

I have tested this with druntime and Phobos, by adding a `printf` when the storage class changes in `target.d` and found only one occurrence left, in Phobos: https://github.com/dlang/phobos/pull/7594
This means that users will be able to use `-preview=in` without needing a special Phobos / druntime, because the mangling won't change (https://github.com/dlang/dmd/pull/11474) and the ABI won't change. If it inadvertently happen in the future, thanks to `in` being mangled differently whether it really means `in` or `in ref`, it will be caught.

When I enabled it by default, it triggered a cascade of errors, because of the `in ref` error. So timeline-wise, I'd like to do the following:
- [ ] Write the specs (IN PROGRESS);
- [ ] Wait for this to be released (v2.094.0);
- [ ] Adapt a few projects;
- [ ] Deprecated `in ref` (hopefully v2.095.0, potentially v2.096.0);
- [ ] Wait for the end of the deprecation period (v2.105.0 at the earliest);
- [ ] Fix druntime to not show `ref` in demangled output, partially reverting https://github.com/dlang/druntime/pull/3188
- [ ] Make `-preview=in` the default;
- [ ] Adapt druntime / Phobos to use it (e.g. for `-checkaction=context`);

**Comment by Geod24** [comment]
@atilaneves : When this was first proposed over 4 months ago, you were against it (https://github.com/dlang/dmd/pull/11000#issuecomment-608415603).
Considering the new semantics, and all the discussions that were brought up here (including [Herb Sutter's explanation on why C++ should have it](https://youtu.be/qx22oxlQmKc?t=1258)), how do you feel about this now ?

@WalterBright @andralex : It would be great to have your opinion on this.

I think we have tons of empirical data to show that there is community support for it, either here, in [the forum thread](https://forum.dlang.org/post/bbycdlmqxkhbucevudgy@forum.dlang.org)), or just by looking how many people use `in` and `in ref` in their projects.

**Comment by Geod24** [comment]
Note: Here is what happen when I turned it on by default: https://buildkite.com/dlang/dmd/builds/15037#56de4045-a8b4-45dc-9ea7-83f437066ddf

**Comment by nordlow** [comment]
I'm praying for this to get merged.

**Comment by kinke** [comment]
> Wait for this to be released (v2.094.0)

I think that's too ambitious; the `Target` hook still needs work to account for the different ABIs (argTypes etc. as mentioned above), and then there'll probably be more ABI changes in druntime/Phobos (especially for Win64).

**Comment by Geod24** [comment]
> (especially for Win64, e.g., `in real` will be a ref for DMD).

I knew I was forgetting something :) Indeed, floating point values get passed in different register so they should be handled differently. And if a struct is made of 3 doubles, do we currently pass it in 3 registers or on the stack ?

TBH, regarding the best ABI, I'd appreciate if you (or people more familiar with calling conventions) could follow up on this. I don't think, at the moment, there are cases where things are passed by value which should be passed by `ref`, it's usually the other way around, and that's on purpose, as from a maintainer PoV, I think it's fine to do adjustments later. Going from `ref` to value has very little chance to break code, while the other way around would, since it would disallow some more common covariance conversion pattern.

**Comment by kinke** [comment]
An example hopefully illustrating why the target ABI is of crucial importance here:
```D
void byRef(const scope ref real);
void byVal(const scope real);

void foo(real param)
{
    byRef(param); // fine, passing a ref to the `param` argument

    // The Win64 ABI requires x87 real to be passed by ref under the hood.
    // So with high-level by-value passing, this requires the caller to make a copy
    // and pass that one by ref. This is exactly one of the things we want to avoid
    // with optimal `in` semantics, passing a ref to the original argument directly
    // instead.
    byVal(param); // on Win64, basically lowered to: `(auto copy = param; byVal(&copy);)`
}
```

**Comment by Geod24** [comment]
> // The Win64 ABI requires x87 real to be passed by ref under the hood.

Is that also true for win32 ?

**Comment by kinke** [comment]
Nope, almost everything is pushed onto the stack for 32-bit x86 calling conventions. Meaning that we get to decide whether to push a 4-bytes pointer or a 10-bytes value.

**Comment by Geod24** [comment]
Thanks @kinke . I made `real` pass-by-ref all the time at the moment.

Another question that popped up: Should delegates infer their storage class ? Currently they (mostly) don't:
```D
void main () @safe
{
    lazyDg((a) => 42);
    refDg((a) => 42);
    outDg((a) => 42);
    scopeDg((a) => 42);
}

void lazyDg (scope int delegate(ref int) @safe sink) @safe {}
void refDg (scope int delegate(lazy int) @safe sink) @safe {}
void outDg (scope int delegate(out int) @safe sink) @safe {}
void scopeDg (scope int delegate(scope int[]) @safe sink) @safe {}
```

```console
test9423.d(3): Error: function test9423.lazyDg(scope int delegate(ref int) @safe sink) is not callable using argument types (void)
test9423.d(3):        cannot pass argument __lambda1 of type void to parameter scope int delegate(ref int) @safe sink
test9423.d(4): Error: function test9423.refDg(scope int delegate(lazy int) @safe sink) is not callable using argument types (void)
test9423.d(4):        cannot pass argument __lambda2 of type void to parameter scope int delegate(lazy int) @safe sink
test9423.d(5): Error: function test9423.outDg(scope int delegate(out int) @safe sink) is not callable using argument types (void)
test9423.d(5):        cannot pass argument __lambda3 of type void to parameter scope int delegate(out int) @safe sink
```

As you can see, inference is only done for `scope`... There's a discussion about this on [this issue](https://issues.dlang.org/show_bug.cgi?id=9423) where Kenji explains the reasoning. The fix is trivial (a single line), but I'm wondering if anyone had opinion on this. The issue I noticed it is that, obviously, it creates problem with `in`.

**Comment by andralex** [comment]
I think this is good fwiw.

**Comment by Geod24** [comment]
Rebased on master. I emailed Walter, too. Not sure what else I can do here to get feedback.

**Comment by ibuclaw** [comment]
> Rebased on master. I emailed Walter, too. Not sure what else I can do here to get feedback.

Give them 72 hours? :-)

As this is going in master, reverting within a week or two is no problem.

File: src/dmd/target.d
Comment (ibuclaw) [inline src/dmd/target.d:780]:
```suggestion
            else if (global.params.isWindows && p.type.ty == Tfloat80)
```

File: src/dmd/target.d
Comment (ibuclaw) [inline src/dmd/target.d:761]: Needs to be added to target.h, maybe with a cxx-unittest if I was to be stringent about it. :-)

File: src/dmd/target.d
Comment (ibuclaw) [inline src/dmd/target.d:796]: What is this really for?  Passing vectors by reference? Passing large structs by reference?  Why not test whether the type should be returned in memory instead?

**Review by ibuclaw — CHANGES_REQUESTED** [review]
Last nits before pushing over the line.

File: src/dmd/target.d
Comment (Geod24) [reply src/dmd/target.d:796]: Large structs and static arrays

File: src/dmd/target.d
Comment (Geod24) [reply src/dmd/target.d:761]: Since it's not a virtual function and there's no vtable, I don't see if that'd be useful

**Comment by Geod24** [comment]
Updated and rebased on master

File: src/dmd/target.d
Comment (ibuclaw) [reply src/dmd/target.d:761]: I need to implement this in gdc, ditto ldc.

Comment (Geod24) [reply src/dmd/target.d:761]: I added it to `target.h`, just not in `cxx-unittest`

File: src/dmd/target.d
Comment (ibuclaw) [reply src/dmd/target.d:797]: And is passing `in __vector` by reference intentional?

Comment (Geod24) [reply src/dmd/target.d:797]: https://github.com/dlang/dmd/pull/11000#issuecomment-674508079

> TBH, regarding the best ABI, I'd appreciate if you (or people more familiar with calling conventions) could follow up on this.

Comment (ibuclaw) [reply src/dmd/target.d:797]: For dmd, `target.isReturnOnStack` or `target.toArgTypes` would probably be the best gauge for deciding whether to pass by ref or not.  There's no point passing by ref if its going in registers.

Comment (kinke) [reply src/dmd/target.d:797]: That's clear and has been postponed to a follow-up; this PR isn't of much use until then.

Comment (ibuclaw) [reply src/dmd/target.d:797]: For DMD, it should be ok to do:
```
else
{
    auto tup = target.toArgTypes(p.type);
    if (!tup.arguments)
        p.storageClass |= STC.ref_;
}
```
Size alone is a bit too arbitrary a gauge.

Comment (Geod24) [reply src/dmd/target.d]: @ibuclaw : Thanks, done

**Review by ibuclaw — APPROVED** [review]

File: src/dmd/target.d
Comment (kinke) [reply src/dmd/target.d]: As shown by CI, this doesn't work for Win64 at all (returns null), *and* is costly for structs, where the argTypes are cached as mentioned earlier.

**Comment by thewilsonator** [comment]
```
fail_compilation/previewin.d(3): Error: function `previewin.func1(void function(ulong[8]) dg)` is not callable using argument types `(void function(in ulong[8]))`
fail_compilation/previewin.d(3):        cannot pass argument `& func_byRef` of type `void function(in ulong[8])` to parameter `void function(ulong[8]) dg`
fail_compilation/previewin.d(4): Error: function `previewin.func2(void function(ref ulong[8]) dg)` is not callable using argument types `(void function(in ulong[8]))`
fail_compilation/previewin.d(4):        cannot pass argument `& func_byRef` of type `void function(in ulong[8])` to parameter `void function(ref ulong[8]) dg`
fail_compilation/previewin.d(5): Error: function `previewin.func3(void function(ref const(ulong[8])) dg)` is not callable using argument types `(void function(in ulong[8]))`
fail_compilation/previewin.d(5):        cannot pass argument `& func_byRef` of type `void function(in ulong[8])` to parameter `void function(ref const(ulong[8])) dg`
fail_compilation/previewin.d(7): Error: function `previewin.func4(void function(ref uint) dg)` is not callable using argument types `(void function(in uint))`
fail_compilation/previewin.d(7):        cannot pass argument `& func_byValue` of type `void function(in uint)` to parameter `void function(ref uint) dg`
fail_compilation/previewin.d(41): Error: scope variable `arg` assigned to non-scope `myGlobal`
fail_compilation/previewin.d(42): Error: scope variable `arg` assigned to non-scope `myGlobal`
fail_compilation/previewin.d(43): Error: scope variable `arg` may not be returned
fail_compilation/previewin.d(44): Error: scope variable `arg` assigned to `escape` with longer lifetime
fail_compilation/previewin.d(48): Error: returning `arg` escapes a reference to local variable `arg`
```

**Comment by Geod24** [comment]
Great, a Windows-only SEGV creeping in...

**Comment by kinke** [comment]
I told you so: https://github.com/dlang/dmd/pull/11000#discussion_r475541123

**Comment by Geod24** [comment]
Ah I missed that comment. So, any suggestion ?

**Comment by kinke** [comment]
IMO, just leave proper ABI handling to a follow-up.

**Comment by Geod24** [comment]
Works for me, reverted to the size check

**Comment by ibuclaw** [comment]
> IMO, just leave proper ABI handling to a follow-up.

Right, the intent just be fine though.
```
if (auto ts = p.type.isTypeStruct())
{
    if (ts.sym.numArgTypes())
         p.storageClass |= STC.ref_;
}
else if (auto tup = .toArgTypes(p.type))  // bypasses target.toArgTypes()
{
    if (tup.arguments)
         p.storageClass |= STC.ref_;
}
```
Just having a guess though.

File: test/runnable/previewin.d
Comment (ibuclaw) [inline test/runnable/previewin.d:160]: FYI, SPARC 64-bit would pass it by value. (you need a type > 32 bytes :-)

Comment (Geod24) [reply test/runnable/previewin.d]: Good catch

**Comment by Geod24** [comment]
All green again.

**Comment by nordlow** [comment]
Wonderful!

**Comment by PetarKirov** [comment]
@Geod24 that's a great achievement, thanks for pushing it through! @kinke @ibuclaw and everyone else, thank you guys for the help as well!

**Comment by ibuclaw** [comment]
image

**Pull request merged**
